File: CSES_PROBLEM/palidrome_reorder.cpp
#include <bits/stdc++.h>
using namespace std;

class PalindromeReorder {
private:
    string s;
    vector<int> freq;

public:
    PalindromeReorder(const string &str) : s(str), freq(26, 0) {}

    // Method to check feasibility
    bool canFormPalindrome() {
        for (char c : s) freq[c - 'A']++;
        int oddCount = 0;
        for (int f : freq) {
            if (f % 2) oddCount++;
        }
        return oddCount <= 1;
    }

    // Method to build palindrome
    string buildPalindrome() {
        if (!canFormPalindrome()) return "NO SOLUTION";

        string firstHalf, secondHalf;
        char oddChar = '\0';
        int oddFreq = 0;

        for (int i = 0; i < 26; i++) {
            if (freq[i] % 2) {
                oddChar = 'A' + i;
                oddFreq = freq[i];
            }
            firstHalf.append(freq[i] / 2, 'A' + i);
        }

        secondHalf = firstHalf;
        reverse(secondHalf.begin(), secondHalf.end());

        if (oddChar != '\0')
            return firstHalf + string(oddFreq, oddChar) + secondHalf;
        else
            return firstHalf + secondHalf;
    }
};

int main() {
    string input;
    cin >> input;

    PalindromeReorder solver(input);
    cout << solver.buildPalindrome() << "\n";

    return 0;
}

File: CSES_PROBLEM/palidrome_reorder.py
from collections import Counter

class PalindromeReorder:
    def __init__(self, s: str):
        self.s = s
        self.freq = Counter(s)

    def build_palindrome(self) -> str:
        # Find odd count characters
        odd_chars = [ch for ch, count in self.freq.items() if count % 2 == 1]

        if len(odd_chars) > 1:
            return "NO SOLUTION"

        first_half = []
        middle = ""

        for ch in sorted(self.freq.keys()):  # sorted for stable output
            count = self.freq[ch]
            if count % 2 == 1:
                middle = ch * count
            first_half.append(ch * (count // 2))

        first_half_str = "".join(first_half)
        return first_half_str + middle + first_half_str[::-1]


if __name__ == "__main__":
    s = input().strip()
    solver = PalindromeReorder(s)
    print(solver.build_palindrome())
